src/main/kotlin/no/nav/syfo/consumer/azuread/v2/AzureAdV2TokenConsumer.kt

```diff
@@ -8,32 +9,77 @@ import org.springframework.util.LinkedMultiValueMap
     @Value("\${azure.app.client.secret}") private val azureAppClientSecret: String,
-    @Value("\${azure.openid.config.token.endpoint}") private val azureTokenEndpoint: String
+    @Value("\${azure.openid.config.token.endpoint}") private val azureTokenEndpoint: String,
+    private val metric: Metrikk,
 ) {
-    fun getToken(
+    fun getOboToken(
         scopeClientId: String,
-        token: String? = null
+        token: String,
+        veilederId: String,
+        azp: String,
     ): String {
-        try {
-            val requestEntity = token?.let { onBehalfOfRequestEntity(scopeClientId, token) } ?: systemTokenRequestEntity(scopeClientId)
-            val response = restTemplateWithProxy.exchange(
-                azureTokenEndpoint,
-                HttpMethod.POST,
-                requestEntity,
-                AzureAdV2TokenResponse::class.java
-            )
-            val tokenResponse = response.body!!
+        val keyForTokenCache = "$veilederId-$azp-$scopeClientId"
+        val cachedToken = tokenCache[keyForTokenCache]
+        return if (cachedToken == null || cachedToken.isExpired()) {
+            try {
+                val requestEntity = onBehalfOfRequestEntity(scopeClientId, token)
+                val response = restTemplateWithProxy.exchange(
+                    azureTokenEndpoint,
+                    HttpMethod.POST,
+                    requestEntity,
+                    AzureAdV2TokenResponse::class.java
+                )
+                val tokenResponse = response.body!!
+                val azureAdV2Token = tokenResponse.toAzureAdV2Token()
+                tokenCache[keyForTokenCache] = azureAdV2Token
+                metric.countEvent(CALL_AZUREAD_CACHE_MISS)
+                log.info("Cache miss for key $keyForTokenCache")
+                azureAdV2Token.accessToken
+            } catch (e: RestClientResponseException) {
+                log.error("Call to get AzureADV2Token from AzureAD for scope: $scopeClientId with status: ${e.rawStatusCode} and message: ${e.responseBodyAsString}", e)
+                throw e
+            }
+        } else {
+            metric.countEvent(CALL_AZUREAD_CACHE_HIT)
+            log.info("Cache hit for key $keyForTokenCache")
+            cachedToken.accessToken
+        }
+    }
 
-            return tokenResponse.toAzureAdV2Token().accessToken
-        } catch (e: RestClientResponseException) {
-            log.error("Call to get AzureADV2Token from AzureAD for scope: $scopeClientId with status: ${e.rawStatusCode} and message: ${e.responseBodyAsString}", e)
-            throw e
+    fun getSystemToken(
+        scopeClientId: String,
+    ): String {
+        val cachedToken = tokenCache[scopeClientId]
+        return if (cachedToken == null || cachedToken.isExpired()) {
+            try {
+                val requestEntity = systemTokenRequestEntity(scopeClientId)
+                val response = restTemplateWithProxy.exchange(
```

Comment:
Kan evt refaktorere ut henting av token (linje 63-74/33-44) for å unngå duplisering her.

Reply:
Fikset nå